src/pixelator/graph/backends/implementations.py

```diff
@@ -428,9 +517,118 @@ def __init__(
+    ) -> Union[nx.Graph, nx.MultiGraph]:
+        unique_markers = set(df.unique("marker").collect()["marker"].to_list())
+        initial_marker_dict = {marker: 0 for marker in unique_markers}
+
+        g: nx.Graph = nx.empty_graph(0, create_using)
+
+        for row in (
+            df.select(["upia", "upib", "marker"])
+            .collect(streaming=True)
+            .iter_rows(named=False, buffer_size=1000)
+        ):
+            # We are duplicating code here, since it gives
+            # a performance boost of roughly 10% here.
+            # Which isn't that much, but will add up when
+            # we have large edge lists.
+            node_1 = row[0]
+            node_2 = row[1]
+            marker = row[2]
+
+            existing_node_1 = g.nodes.get(node_1)
+            if existing_node_1:
+                existing_node_1["markers"][marker] += 1
+            else:
+                marker_dict = initial_marker_dict.copy()
+                marker_dict[marker] += 1
+                g.add_node(node_1, markers=marker_dict)
+
+            existing_node_2 = g.nodes.get(node_2)
+            if existing_node_2:
+                existing_node_2["markers"][marker] += 1
+            else:
+                marker_dict = initial_marker_dict.copy()
+                marker_dict[marker] += 1
+                g.add_node(node_2, markers=marker_dict)
+
+            g.add_edge(node_1, node_2)
+
+        return g
+
+    @staticmethod
+    def _add_node_attributes(
+        graph: Union[nx.Graph, nx.MultiGraph], a_nodes: set[str]
+    ) -> None:
+        node_names = {node: node for node in graph.nodes()}
+        pixel_type = {node: "A" if node in a_nodes else "B" for node in graph.nodes()}
+        type_ = {node: node in a_nodes for node in graph.nodes()}
+        nx.set_node_attributes(graph, node_names, "name")
+        nx.set_node_attributes(graph, pixel_type, "pixel_type")
+        nx.set_node_attributes(graph, type_, "type")
+
+    @staticmethod
+    def _project_on_a_nodes(
+        graph: Union[nx.Graph, nx.MultiGraph], a_nodes: set[str]
+    ) -> Union[nx.Graph, nx.MultiGraph]:
+        if isinstance(graph, nx.MultiGraph):
+            warnings.warn(
+                "Using `use_full_bipartite=True` together with `simplify=False` "
+                "will still implicitly simplify the graph."
+            )
+            graph = nx.Graph(graph)
+
+        return bipartite.projected_graph(graph, a_nodes)
+
+    @staticmethod
+    def _build_graph_with_marker_counts(
+        edgelist: pl.LazyFrame, simplify: bool, use_full_bipartite: bool
+    ) -> Union[nx.Graph, nx.MultiGraph]:
+        graph = NetworkXGraphBackend._build_graph_with_node_counts_from_edgelist(
+            edgelist,
+            create_using=nx.Graph if simplify else nx.MultiGraph,
+        )
+        a_nodes = set(edgelist.select(["upia"]).unique().collect()["upia"].to_list())
+        NetworkXGraphBackend._add_node_attributes(graph, a_nodes)
+        if use_full_bipartite:
```

Comment:
Same here as below, return the graph if we don't need to compute A-node projection

Reply:
This needs to be in place as it is needed for plotting. It is also tested as is.